File: projects/18-elevator-escalator-fleet-monitoring/backend/dashboard/lambdaHandler.js
"use strict";

const { buildClients } = require("./awsClients");
const { buildDeps, buildRouter } = require("./server");

const CORS = {
  "Access-Control-Allow-Origin": "*",
  "Access-Control-Allow-Methods": "GET,OPTIONS",
  "Access-Control-Allow-Headers": "*",
};

let warmRouter = null;
function routerFor(injected) {
  // Lambda invokes (event, context, callback); accept a third arg as clients only when it carries a doc client.
  if (injected && injected.doc) return buildRouter(buildDeps(injected));
  if (!warmRouter) warmRouter = buildRouter(buildDeps(buildClients()));
  return warmRouter;
}

function captureRes() {
  const captured = { status: 200, body: "" };
  return {
    captured,
    writeHead(status) { captured.status = status; return this; },
    end(text) { captured.body = text == null ? "" : String(text); },
  };
}

function reply(status, body) {
  return {
    statusCode: status,
    headers: { "Content-Type": "application/json", ...CORS },
    body: typeof body === "string" ? body : JSON.stringify(body),
  };
}

async function handler(event, _context, injectedClients) {
  const method = event.httpMethod || "GET";
  const rawPath = event.path || "/";
  if (method === "OPTIONS") return { statusCode: 200, headers: CORS, body: "" };

  const query = new URLSearchParams(event.queryStringParameters || {}).toString();
  const url = new URL(rawPath + (query ? `?${query}` : ""), "http://localhost");
  const res = captureRes();
  try {
    const matched = await routerFor(injectedClients).dispatch(method, url.pathname, {}, res, { url });
    if (!matched) return reply(404, { error: "not found" });
    return { statusCode: res.captured.status, headers: { "Content-Type": "application/json", ...CORS }, body: res.captured.body };
  } catch (err) {
    return reply(500, { error: err.message || "internal error" });
  }
}

module.exports = { handler, routerFor };

File: projects/27-river-flood-monitoring/backend/dashboard/lambda_handler.py
"""API Gateway entry point: dispatch the proxy event's path to the shared view functions and add CORS."""
import json

import views

CORS = {
    "Access-Control-Allow-Origin": "*",
    "Access-Control-Allow-Methods": "GET,OPTIONS",
    "Access-Control-Allow-Headers": "*",
}


def _normalise(path):
    return path[:-1] if len(path) > 1 and path.endswith("/") else path


def _response(status, body):
    return {"statusCode": status, "headers": {"Content-Type": "application/json", **CORS}, "body": json.dumps(body)}


def lambda_handler(event, context):
    if event.get("httpMethod") == "OPTIONS":
        return {"statusCode": 200, "headers": dict(CORS), "body": ""}
    view = views.ROUTES.get(_normalise(event.get("path", "/")))
    if view is None:
        return _response(404, {"error": "not found"})
    status, body = view(event.get("queryStringParameters") or {})
    return _response(status, body)
